src/NULS/Address.cpp

```diff
@@ -0,0 +1,133 @@
+// Copyright © 2017-2019 Trust Wallet.
+//
+// This file is part of Trust. The full Trust copyright notice, including
+// terms governing use, modification, and redistribution, is contained in the
+// file LICENSE at the root of the source code distribution tree.
+#include "Address.h"
+
+#include <TrezorCrypto/ecdsa.h>
+
+#include "../Base58.h"
+#include "../BinaryCoding.h"
+#include "../HexCoding.h"
+
+using namespace TW::NULS;
+
+bool Address::isValid(const std::string& string)
+{
+    if (string.empty()) {
+        return false;
+    }
+
+    Data decoded = TW::Base58::bitcoin.decode(string);
+    if (decoded.size()!=Address::addressSize) {
+        return false;
+    }
+
+    // Check Xor
+    uint8_t checkSum = 0x00;
+    for (int i = 0; i<23; ++i) {
+        checkSum ^= decoded[i];
+    }
+
+    return decoded[23]==checkSum;
+}
+
+Address::Address(const std::string& string)
+{
+    if (!isValid(string)) {
+        return;
+    }
+
+    Data decoded = TW::Base58::bitcoin.decode(string);
+    std::copy(decoded.begin(), decoded.end(), bytes.begin());
+}
+
+Address::Address(const TW::PublicKey& publicKey)
+{
+    // Main-Net chainID
+    bytes[0] = 0x04;
+    bytes[1] = 0x23;
+    // Address Type
+    bytes[2] = 0x01;
+
+    ecdsa_get_pubkeyhash(publicKey.bytes.data(), HASHER_SHA2_RIPEMD, bytes.begin()+3);
+
+    // Calc chechsum
+    uint8_t checkSum = 0x00;
+    for (int i = 0; i<23; ++i) {
+        checkSum ^= bytes[i];
+    }
+    bytes[23] = checkSum;
+}
+
+Address::Address(const std::vector<uint8_t>& data)
+{
+    if (data.size()!=Address::addressSize) {
+        throw std::invalid_argument("Invalid address data");
+    }
+
+    std::copy(data.begin(), data.end(), bytes.begin());
+}
+
+std::string Address::string() const
+{
+    return TW::Base58::bitcoin.encode(bytes.begin(), bytes.end());
+}
+
+uint16_t Address::chainID() const
+{
+    return decode16LE(bytes.data());
+}
+
+uint8_t Address::type() const
+{
+    return bytes[2];
+}
+
+bool Address::isValid() const
+{
+    if (chainID()!=8964) {
+        return false;
+    }
+    if (type()!=0x01 && type()!=0x02) {
```

Comment:
let's add space between operators `!=`, `==`, `<`
```suggestion
    if (type() != 0x01 && type() != 0x02) {
```